### ISSUE TITLE
Issue on use_gcn_feat

### ISSUE THREAD

**Issue by keeprunningcelin** [yl-1993/learn-to-cluster#42]
Hi Yang Lei, I met an issue on setting in configuration.In my project, the cluster data is vehicle data which is using for vehicle reid task.
In cfg_test_gcnv_*.py，have use_gcn_feat=True by default. After I train gcn-v model, I test gcn-v model. Here are the results.

Model | KNN K value | Metric | Presicion | Recall | Pairwise F-score
GCN-V | 60 | Snbr_t=0.8 | 0.8868 | 0.8041 | 0.8435
GCN-V | 60 | Snbr_t=0.85 | 0.9569 | 0.7799 | 0.8594
GCN-V | 60 | Snbr_t=0.9 | 0.9861 | 0.7001 | 0.8188
GCN-V | 60 | SFnbr_t=0.85 | 0.6536 | 0.7563 | 0.7012
GCN-V | 60 | SFnbr_t=0.9 | 0.8632 | 0.7366 | 0.7949
GCN-V | 60 | SFnbr_t=0.95 | 0.9688 | 0.6581 | 0.7838

SFnber indicates that using the transformed features to rebuild the affinity graph achieves relatively lower performance.And Two definitions are sensitive to the threshold.
I'd like to know how to set the threshold in test configuration for  gcn-v and gcn-e models in detail.
Thanks.

**Comment by yl-1993** [comment]
Hi @keeprunningcelin, thanks for trying our algorithm on your datasets.
(1) For the transformed features, our empirical results show that the performance of transformed features usually depends on the training. If the training goes well, it is more likely that it receives performance gain. Besides, it may depend on the dimension of transformed features, but it seems that `512` works well on our settings where input feature dimension is `256`. I would suggest to first check the loss of training.
(2) For the threshold, it can be determined by constructing a small validation set which contains both positive pairs and negative pairs. We can compute the similarity scores for all pairs and find a threshold to maximize the accuracy on the set.
(3) If GCN-E is based on the rebuilt graph, then the threshold is likely to be the same as GCN-V. For example, `tau=0.85` is shared between [gcnv_fashion](https://github.com/yl-1993/learn-to-cluster/blob/master/vegcn/configs/cfg_test_gcnv_fashion.py#L33) and [gcne_fashion](https://github.com/yl-1993/learn-to-cluster/blob/master/vegcn/configs/cfg_test_gcne_fashion.py#L33), and `tau=0.8` is shared between [gcnv_ms1m](https://github.com/yl-1993/learn-to-cluster/blob/master/vegcn/configs/cfg_test_gcnv_ms1m.py#L33) and [gcne_ms1m](https://github.com/yl-1993/learn-to-cluster/blob/master/vegcn/configs/cfg_test_gcne_ms1m.py#L33).

**Comment by keeprunningcelin** [comment]
Hi @yl-1993,Thanks for your reply! 
(1) input feature dimension in my data is 512, and nhid is 512 by default. Maybe I need to dajust nhid value to retrain?
(2)I train GCN-E which is not based on the rebuilt graph. I compare k = 60 and k=120 in GCN-E.
k=120 in GCN-E brings more performance gain. k = 60 in GCN-E and  k=60 in GCN-V have similar performance.
(3)If GCN-E is not based on the rebuilt graph, "tau" in GCN-E is likely to be the same as "tau_0" in GCN-V?
Thanks.

**Comment by yl-1993** [comment]
Hi @keeprunningcelin,
(1) You can try to retrain the model if you have time. In addition, you can try more fancy GCN architectures for GCN-V. For example, residual GCN adds or concats the input features to the transformed features.
(2) Yes, it works as expected. As stated in https://github.com/yl-1993/learn-to-cluster/issues/39#issuecomment-619783907, increasing k for GCN-E usually receives more performance gain.
(3) Yes. You're right.
